# Problem Set 4

In [2]:
import numpy as np
import pandas as pd
import random

import sklearn.model_selection as model_selection
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.model_selection import train_test_split

from sklearn.model_selection import cross_val_predict, cross_validate
from sklearn import linear_model


import matplotlib.pyplot as plt
%matplotlib inline

## Problem A: Import 10000 records from the NYC yellow cab data set
Import 100 records from the NYC yellow cab data set. We're going to be looking at the relationship between trip distance and tip amount. Are passengers fwho travel further tipping better than others?

Our first task is to import and clean the data. Import the first 10000 rows of the Jan 2017 yellow cab NYC trip data into a pandas data frame. Create a view on trip_distance and tip_amount. Plot the trip_distance vs. tip amount in a scatter plot.

The data set can be found here:
<http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml>

In [11]:
# IMPLEMENT ME (Please make sure you name the dataframe as 'df1' and don't modify 'df1' in later parts.)
# (Read your data from "yellow_tripdata_2017-01.csv".)

df1 = pd.read_csv("yellow_tripdata_2017-01.csv", nrows = 10000)
df1.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2017-01-09 11:13:28   2017-01-09 11:25:45                1   
1         1  2017-01-09 11:32:27   2017-01-09 11:36:01                1   
2         1  2017-01-09 11:38:20   2017-01-09 11:42:05                1   
3         1  2017-01-09 11:52:13   2017-01-09 11:57:36                1   
4         2  2017-01-01 00:00:00   2017-01-01 00:00:00                1   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.30           1                  N           263           161   
1           0.90           1                  N           186           234   
2           1.10           1                  N           164           161   
3           1.10           1                  N           236            75   
4           0.02           2                  N           249           234   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         12.5    0.0      0.5        2.00           0.0   
1             1          5.0    0.0      0.5        1.45           0.0   
2             1          5.5    0.0      0.5        1.00           0.0   
3             1          6.0    0.0      0.5        1.70           0.0   
4             2         52.0    0.0      0.5        0.00           0.0   

   improvement_surcharge  total_amount  
0                    0.3         15.30  
1                    0.3          7.25  
2                    0.3          7.30  
3                    0.3          8.50  
4                    0.3         52.80

## Problem B: Remove the people who tipped nothing
People who tipped nothing aren't very useful for our needs. We are interesteded in how much they tipped when they were generous enough to do so. Similarly, we don't care about people that tipped for very short rides. Remove all records where the trip distance was less than 0.5 miles. Create a data frame with the records of zero (or less) tips removed and plot the relationship between the trip distance and the tip amount. Can  the resulting data frame 'df'.

In [16]:
# IMPLEMENT ME (Please make sure you name the dataframe as 'df2' and don't modify 'df2' in later parts.)

df2 = pd.DataFrame(columns = df1.columns)
dev_count=  0
for i, trip_info in df1.iterrows():
    if(trip_info["tip_amount"] > 0) and trip_info["trip_distance"] >= 0.5:
        df2 = df2.append(trip_info)
    
df2.head()
    

VendorID tpep_pickup_datetime tpep_dropoff_datetime passenger_count  \
0        1  2017-01-09 11:13:28   2017-01-09 11:25:45               1   
1        1  2017-01-09 11:32:27   2017-01-09 11:36:01               1   
2        1  2017-01-09 11:38:20   2017-01-09 11:42:05               1   
3        1  2017-01-09 11:52:13   2017-01-09 11:57:36               1   
6        2  2017-01-01 00:00:02   2017-01-01 00:39:22               4   

   trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           3.30          1                  N          263          161   
1           0.90          1                  N          186          234   
2           1.10          1                  N          164          161   
3           1.10          1                  N          236           75   
6           7.75          1                  N          186           36   

  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0            1         12.5    0.0      0.5        2.00           0.0   
1            1          5.0    0.0      0.5        1.45           0.0   
2            1          5.5    0.0      0.5        1.00           0.0   
3            1          6.0    0.0      0.5        1.70           0.0   
6            1         22.0    0.5      0.5        4.66           0.0   

   improvement_surcharge  total_amount  
0                    0.3         15.30  
1                    0.3          7.25  
2                    0.3          7.30  
3                    0.3          8.50  
6                    0.3         27.96

# Problem C: Fit a linear regression to the resulting trip_distance vs. tip_amount data

In [17]:
X = df2['trip_distance'].values.reshape(-1,1)
y = df2['tip_amount']

In [19]:
# IMPLEMENT ME

regression = LinearRegression().fit(X,y)

## Problem D: Look at the regression coefficients
Print the regression coefficients (intercept and slope) and in a markdown cell comment about their values. Do they make sense? What do they tell you?

In [23]:
# IMPLEMENT ME (Save the value of intercept as 'a' and the value of slope as 'b'.
# Make sure you only save the value of slope as 'b' instead of an array.)
a = regression.intercept_
b = regression.coef_
print(a)
print(b)

0.869890453417975
[0.5422813]


The intercept is 0.86989 and the slope is 0.5422813. The intercept tells us that the standard tip amount is \$$0.87 , and for every additional mile on the journey, the tip amount is expected to increase $0.54

## Problem E: Evaluate the model
Just because the model fit the data set doesn't mean it's very predictive. Run a 10-fold cross-validation and compute the average mean absolute deviation. Show the result. In a markdown cell comment on what that number means in terms of predictive accuracy.

In [24]:
random.seed(5)
# IMPLEMENT ME (Save the computed average mean absolute deviation as 'c'.)

In [37]:
scores = cross_validate(regression, X, y, cv=10, scoring = "neg_median_absolute_error", return_train_score = False)
print(scores)
np.mean(-1*scores['test_score'])

{'fit_time': array([0.00654507, 0.00201249, 0.00231218, 0.00326991, 0.0021317 ,
       0.00170684, 0.00263572, 0.00267386, 0.00260997, 0.00258088]), 'score_time': array([0.00228167, 0.00166869, 0.00104523, 0.00110602, 0.00095606,
       0.00244641, 0.00136113, 0.00097418, 0.00116467, 0.00136638]), 'test_score': array([-0.35072969, -0.3839472 , -0.44009469, -0.39229226, -0.4058347 ,
       -0.43614265, -0.3734769 , -0.33614418, -0.32383139, -0.30603031])}


0.37485239694852546

The average mean absolute deviation shows the deviation from the central point of the data, that is, the average amount tipped 